proposals/xxxx-work-graphs.md

```diff
@@ -0,0 +1,535 @@
+  /// memory is finalized. The behavior of writes to the output after this call
+  /// is undefined.
+  ///
+  /// Calls to `OutputComplete` must be in thread group uniform control flow
+  /// otherwise the behavior is undefined.
+  void OutputComplete();
+};
+} // namespace detail
+
+template <typename RecordTy>
+using ThreadNodeOutputRecords = detail::NodeOutputRecordsBase<RecordTy>;
+
+template <typename RecordTy>
+using GroupNodeOutputRecords = detail::NodeOutputRecordsBase<RecordTy>;
+```
+
+##### Entry Parameter Attributes
+
+> Consistent with HLSL grammar attribute names are case-insensitive. Any
+> attributes that take string arguments, the string argument values are
+> case-sensitive.
+
+###### **`[MaxRecords(<count>)]`**
+Applies to node inputs in coalescing launch nodes or outputs for any node mode.
+
+Required for node inputs for coalescing launch nodes, this attribute restricts
+the maximum number of records per thread group. Implementations are not required
+to fill to the specified maximum.
+
+When applied to node outputs, this attribute restricts the maximum number of
+records produced to the output. When applied to a `NodeOutputArray`, the maximum
+applies as the sum of all records across the output array, not per-node.
+
+Node outputs require either the `MaxRecords` or `MaxRecordsSharedWith`
+attribute.
+
+###### **`[MaxRecordsSharedWith("<name>")]`**
+This attribute applies to node outputs. The named node must have the
+`MaxRecords` attribute applied to its entry. This attribute and the `MaxRecords`
+attribute are mutually exclusive.
+
+The node that this attribute is applied to shares a maximum record allocation
+with the named node.
+
+Node outputs require either the `MaxRecords` or `MaxRecordsSharedWith`
+attribute.
+
+###### **`[NodeID("<name>", <index> = 0)]`**
+This attribute applies to output nodes and defines the name and index of the
+output node. If not provided, the default index is 0.
+
+If this attribute is not present on an output, the default node ID for the
+output is the name of the parameter, and the index is the default index (0).
+
+###### **`[AllowSparseNodes]`**
+This attribute applies to outputs and allows the work graph to be created even
+if there is not a node defined for the specified output.
+
+###### **`[NodeArraySize(<count>)]`**
+Specifies the output array size for `NodeOutputArray` or `EmptyNodeOutputArray`
+objects.
+
+### New Built-in Functions
+
+#### GetRemainingRecursionLevels
+
+```c++
+/// @brief Returns the number of recursion levels remaining against the declared
+/// `NodeMaxRecursionDepth`.
+///
+/// Returns 0 for leaf nodes and if the current node is not recursive.
+uint GetRemainingRecursionLevels();
+```
+
+For nodes that recurse, the `GetRemainingRecursionLevels()` function returns the
+number of remaining recursion levels before reaching the node's maximum
+recursion depth.
+
+#### Barrier
+
+```c++
+enum class MEMORY_TYPE_FLAG : uint {
```

Comment:
In HLSL, `MEMORY_TYPE_FLAG` and `SEMANTIC_FLAG` are not `enum class`, they are like `enum` instead.  They do not require `MEMORY_TYPE_FLAG::` for use.